lib/deref.js

```diff
@@ -118,16 +119,26 @@ module.exports = {
     }
 
     if (schema.anyOf) {
-      return this.resolveRefs(schema.anyOf[0], parameterSourceOption, components, schemaResolutionCache, stack);
+      return this.resolveRefs(schema.anyOf[0], parameterSourceOption, components, schemaResolutionCache, stack,
+        seenRef);
     }
     if (schema.oneOf) {
-      return this.resolveRefs(schema.oneOf[0], parameterSourceOption, components, schemaResolutionCache, stack);
+      return this.resolveRefs(schema.oneOf[0], parameterSourceOption, components, schemaResolutionCache, stack,
+        seenRef);
     }
     if (schema.allOf) {
-      return this.resolveAllOf(schema.allOf, parameterSourceOption, components, schemaResolutionCache, stack);
+      return this.resolveAllOf(schema.allOf, parameterSourceOption, components, schemaResolutionCache, stack, seenRef);
     }
     if (schema.$ref) {
       let refKey = schema.$ref;
+
+      // if this reference is seen before, ignore and move on.
+      if (seenRef[refKey] < stack) {
```

Comment:
why `< stack` and not `< stackLimit`?

Reply:
We don't have anything like a stackLimit yet. We still have hard coded 20. 
And, I save the stack level first time the reference was encountered, and it is a circular reference only when the next reference occurs at a level more than that. 

Non-circular: 
```
b: 
  a: $ref - stack level 2
  a: $ref - stack level 2

```
Circular: 
```
b: 
   a: $ref - stack level 2
   c: 
      a: $ref - stack level 3
```

Reply:
@abhijitkane